File: security-svc/src/test/java/com/chilly/security_svc/containers/PostgresTestContainer.java
package com.chilly.security_svc.containers;

import org.testcontainers.containers.PostgreSQLContainer;

public class PostgresTestContainer extends PostgreSQLContainer<PostgresTestContainer> {

    public static final String IMAGE_VERSION = "postgres:alpine";
    public static final String DATABASE_NAME = "test";
    private static PostgresTestContainer container;

    public PostgresTestContainer() {
        super(IMAGE_VERSION);
    }

    public static PostgresTestContainer getInstance() {
        if (container == null) {
            container = new PostgresTestContainer().withDatabaseName(DATABASE_NAME);
        }
        return container;
    }

    @Override
    public void start() {
        super.start();
        System.setProperty("DB_URL", container.getJdbcUrl());
        System.setProperty("DB_USER", container.getUsername());
        System.setProperty("DB_PASSWORD", container.getPassword());
    }

    @Override
    public void stop() {}
}

File: places-svc/src/test/kotlin/com/chilly/places_svc/containers/PostgresTestContainer.kt
package com.chilly.places_svc.containers

import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.utility.DockerImageName

class PostgresTestContainer private constructor() : PostgreSQLContainer<PostgresTestContainer>(IMAGE_VERSION) {

    override fun start() {
        super.start()
        container?.let {
            System.setProperty("DB_URL", it.jdbcUrl)
            System.setProperty("DB_USER", it.username)
            System.setProperty("DB_PASSWORD", it.password)
        }
    }

    override fun stop() = Unit

    companion object {
        private val IMAGE_VERSION = DockerImageName.parse("postgis/postgis:17-3.5-alpine")
            .asCompatibleSubstituteFor("postgres")
        private const val DATABASE_NAME = "test"

        private var container: PostgresTestContainer? = null

        operator fun invoke(): PostgresTestContainer {
            return container ?: PostgresTestContainer()
                .withDatabaseName(DATABASE_NAME)
                .withReuse(true)
                .also { container = it }
        }
    }
}
